linera-views/src/backends/lru_caching.rs

```diff
@@ -50,54 +298,265 @@ static NUM_CACHE_SUCCESS: LazyLock<IntCounterVec> = LazyLock::new(|| {
+        match self.map_value.entry(key) {
             btree_map::Entry::Occupied(mut entry) => {
-                entry.insert(value);
-                // Put it on first position for LRU
-                self.queue.remove(&key);
-                self.queue.insert(key, ());
+                entry.insert(cache_entry);
+                self.remove_from_queue(&full_key);
             }
             btree_map::Entry::Vacant(entry) => {
-                entry.insert(value);
-                self.queue.insert(key, ());
-                if self.queue.len() > self.max_cache_size {
-                    let Some(value) = self.queue.pop_front() else {
-                        unreachable!()
-                    };
-                    self.map.remove(&value.0);
-                }
+                entry.insert(cache_entry);
+            }
+        }
+        self.insert_queue(full_key, cache_size);
+    }
+
+    /// Invalidates corresponding find entries
+    pub fn correct_find_entry(&mut self, key: &[u8], new_value: Option<Vec<u8>>) {
+        let lower_bound = self.get_lower_bound_mut(key);
+        if let Some((lower_bound, cache_entry)) = lower_bound {
+            let key_red = &key[lower_bound.len()..];
+            cache_entry.update_cache_entry(key_red, new_value);
+        }
+    }
+
+    /// Inserts a read_value entry into the cache.
+    pub fn insert_read_value(&mut self, key: Vec<u8>, value: &Option<Vec<u8>>) {
+        let cache_entry = ValueCacheEntry::from_read_value(value);
+        self.insert_value(key, cache_entry)
+    }
+
+    /// Inserts a read_value entry into the cache.
+    pub fn insert_contains_key(&mut self, key: Vec<u8>, result: bool) {
+        let cache_entry = ValueCacheEntry::from_contains_key(result);
+        self.insert_value(key, cache_entry)
+    }
+
+    /// Inserts a find entry into the cache.
+    pub fn insert_find(&mut self, key_prefix: Vec<u8>, cache_entry: FindCacheEntry) {
+        let cache_size = cache_entry.size() + key_prefix.len();
+        if cache_size > self.lru_cache_policy.max_cache_size {
+            // Inserting that entry would lead to complete clearing of the cache
+            // which is counter productive
+            return;
+        }
+        let full_prefix = (key_prefix.clone(), CacheEntry::Find);
+        let keys = self
+            .map_find
+            .range(get_interval(key_prefix.clone()))
+            .map(|(x, _)| x.clone())
+            .collect::<Vec<_>>();
+        for key in keys {
+            self.map_find.remove(&key);
+            let full_key = (key, CacheEntry::Find);
+            self.remove_from_queue(&full_key);
+        }
+        match self.map_find.entry(key_prefix.clone()) {
+            btree_map::Entry::Occupied(mut entry) => {
+                entry.insert(cache_entry);
+                self.remove_from_queue(&full_prefix);
+            }
+            btree_map::Entry::Vacant(entry) => {
+                entry.insert(cache_entry);
             }
```

Comment:
If I'm reading this correctly, we are not removing any `map_value` entries that are redundant now? So those would be duplicated in memory.

On the other hand, if we _do_ delete them, their lifetime in the cache is now tied to the lifetime of the whole collection.

This is why I'm not convinced by the whole idea of having entire `find_key_values` results as a single entry in the LRU cache.